gdal/port/cpl_vsil_hdfs.cpp

```diff
@@ -0,0 +1,320 @@
+      if (poInfo != NULL) {
+        tOffset nSize = poInfo->mSize;
+        hdfsFreeFileInfo(poInfo, 1);
+        return static_cast<vsi_l_offset>(nSize);
+      }
+      return -1;
+    }
+    int Eof() override;
+    int Flush() override;
+    int Close() override;
+};
+
+VSIHdfsHandle::VSIHdfsHandle(hdfsFile _poFile,
+                             hdfsFS _poFilesystem,
+                             const char * pszFilename,
+                             bool _bReadOnly)
+  : poFile(_poFile), poFilesystem(_poFilesystem), oFilename(pszFilename), bReadOnly(_bReadOnly)
+{}
+
+VSIHdfsHandle::~VSIHdfsHandle()
+{
+  this->Close();
+}
+
+int VSIHdfsHandle::Seek(vsi_l_offset nOffset, int nWhence)
+{
+  switch(nWhence) {
+  case SEEK_SET:
+    return hdfsSeek(this->poFilesystem, this->poFile, nOffset);
+  case SEEK_CUR:
+    return hdfsSeek(this->poFilesystem,
+                    this->poFile,
+                    nOffset + this->Tell());
+  case SEEK_END:
+    return hdfsSeek(this->poFilesystem,
+                    this->poFile,
+                    static_cast<tOffset>(this->Length()) - nOffset);
+  default:
+    return -1;
+  }
+}
+
+vsi_l_offset VSIHdfsHandle::Tell()
+{
+  return hdfsTell(this->poFilesystem, this->poFile);
+}
+
+size_t VSIHdfsHandle::Read(void *pBuffer, size_t nSize, size_t nMemb)
+{
+  return hdfsRead(this->poFilesystem, this->poFile, pBuffer, nSize * nMemb);
+}
+
+size_t VSIHdfsHandle::Write(const void *pBuffer, size_t nSize, size_t nMemb)
+{
+  CPLError(CE_Failure, CPLE_AppDefined, "HDFS driver is read-only");
+  return -1;
+}
+
+int VSIHdfsHandle::Eof()
+{
+  return (this->Tell() == this->Length());
+}
+
+int VSIHdfsHandle::Flush()
+{
+  return hdfsFlush(this->poFilesystem, this->poFile);
+}
+
+int VSIHdfsHandle::Close()
+{
+  int retval = 0;
+
+  if (this->poFilesystem != nullptr && this->poFile != nullptr)
+    retval = hdfsCloseFile(this->poFilesystem, this->poFile);
+  this->poFile = nullptr;
+  this->poFilesystem = nullptr;
+
+  return retval;
+}
+
+class VSIHdfsFilesystemHandler final : public VSIFilesystemHandler
+{
+  private:
+    CPL_DISALLOW_COPY_ASSIGN(VSIHdfsFilesystemHandler)
+
+    hdfsFS poFilesystem;
+
+  public:
+    VSIHdfsFilesystemHandler();
+    ~VSIHdfsFilesystemHandler() override;
+
+    VSIVirtualHandle *Open(const char *pszFilename,
+                           const char *pszAccess,
+                           bool bSetError ) override;
+    int Stat(const char *pszFilename,
+             VSIStatBufL *pStatBuf,
+             int nFlags) override;
+    int Unlink(const char *pszFilename) override;
+    int Mkdir(const char *pszDirname, long nMode) override;
+    int Rmdir(const char *pszDirname) override;
+    char ** ReadDir(const char *pszDirname) override;
+    int Rename(const char *oldpath, const char *newpath) override;
+
+};
+
+VSIHdfsFilesystemHandler::VSIHdfsFilesystemHandler()
+{
+  this->poFilesystem = hdfsConnect("default", 0);
```

Comment:
what is the overhead of this ? This constructor will be called at GDAL startup when VSI is initialized, so if there's significant time involved, the connection should be moved to the first time a real access (open, stat, etc) is done

Reply:
Done.